### ISSUE TITLE
Mapped type of polymorphic this type, containing a property type condition, fails when used within a method

### ISSUE THREAD

**Issue by yoni-rapoport** [microsoft/TypeScript#56672]
### 🔎 Search Terms

"Mapped type", "polymorphic this type", "property type condition"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Mapped type of polymorphic this type

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.3.2#code/C4TwDgpgBA8gtgS2AMQK4DsDGwEHt0DOAPACoB8UAvFAN5QDaA0lAulANYQi4BmUJUAIYF+TALpQIAD2AR0AExFosOfFAD8UdBABuEAE5QAXFEZiTJcVAC+AKFuYANsJEBZEAGFnBETVtQhKigAcmD7AJ4ACgBKE3gkZWw8QiJgAAsEAgo6fQhgVH02GmshEQx2dFwAdzZhWEQUDCT8YnTMijt-KABzGNougO0qqHcvFxiAOijoicEoAHp5qCrcfXYROulIbAh5cICoNoIpybnFw4JIgCYAZhuATmioAFpJKW3ZeS1cSX19VYI+wCmBawCgUiCxVK9QSTVUKSOZAG4NmCyWwEutweT1eWwgOy+lV+-30gORIMIYJAkJKdQASnkCugSOAIKkMgR6MEeMExEiDlAQKjzqxZLkCDh0N1HNSVmsCABCLp2axAA

### 💻 Code

```ts
type OmitFunctions<T> = { [K in keyof T as T[K] extends Function ? never : K]: T[K] }

class MyClass {
  a = ''

  f(): OmitFunctions<this> { return {} as unknown as OmitFunctions<this> }

  g() {
    new MyClass().f().a // works as expected

    this.f().a // ts(2339) - expected no errors

    const x = {} as OmitFunctions<this>
    x.a // ts(2339) - expected no errors

    const y = {} as ReturnType<this['f']>
    y.a // interestingly works!
  }
}
```

### 🙁 Actual behavior

ts2339 - Property 'a' does not exist on type 'OmitFunctions<this>'.

### 🙂 Expected behavior

No error

**Comment by RyanCavanaugh** [comment]
`OmitFunctions<this>` has to remain deferred to be correct in higher-order instantiations. A derived class of `MyClass` could have a subtype of `a` that *did* get omitted, e.g.:
```ts
type OmitEmptyStrings<T> = { [K in keyof T as T[K] extends "" ? never : K]: T[K] }

class MyClass {
  a = 'foo';

  f(): OmitEmptyStrings<this> { return {} as unknown as OmitEmptyStrings<this> }

  g2() {
    const x = this.f();
    x.a; // alleged: this should be legal,
    // but this is inconsistent with line 21
    return x;
  }
}

class MyDerivedClass extends MyClass {
  a: "" = "";
}

const d = new MyDerivedClass();
d.f().a; // Not legal
```

**Comment by yoni-rapoport** [comment]
@RyanCavanaugh how is that different from other cases of property type modification in a derived class? e.g.:

```ts
class MyClass {
  a = 'foo';

  g2() {
    this.a = '123' // legal
  }
}

class MyDerivedClass extends MyClass {
  a: "" = "";
}

const d = new MyDerivedClass();
d.a = '123'; // Not legal
```

**Comment by yoni-rapoport** [comment]
I think this issue should be re-opened. The current state has a significant negative impact on the use of mapped types in class methods.

**Comment by whzx5byb** [comment]
@yoni-rapoport 
> how is that different from other cases of property type modification in a derived class

The code you posted in https://github.com/microsoft/TypeScript/issues/56672#issuecomment-1848915042 is an well-known unsoundness. Ideally, if a property is writable, it should be treated as invariant, so overriding `a: string` with `a: ''` should be an error. Unfortunately, this unsoundness may never be fixed.

**Comment by yoni-rapoport** [comment]
@whzx5byb 

> The code you posted in [#56672 (comment)](https://github.com/microsoft/TypeScript/issues/56672#issuecomment-1848915042) is an well-known unsoundness. Ideally, if a property is writable, it should be treated as invariant, so overriding `a: string` with `a: ''` should be an error. Unfortunately, this unsoundness may never be fixed.

That's exactly my point. Why is it "working as intended" to have mapped types of `this` not work inside class methods, only to accommodate for scenarios that are erroneous (and I would argue also extremely rare)?

@RyanCavanaugh ?
